### PR TITLE
Add experimental plugin to initiate H2 Server push for preload links.

### PR DESCRIPTION
**Pull request by calavera**
This plugin parses the response Link header to extract assets to preload.
Then an H2 Server Push is initiated for those assets.

I appreciate feedback, specially around improvements to parse the Link header in a more efficient way.

/cc @maskit 

Signed-off-by: David Calavera <david.calavera@gmail.com>

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci]

**Comment by zwoop** [comment]
I'm not an expert in this area of the code, so I can't really comment on the specific. But an immediate concern here would be that we are at risk of pushing the same pieces over and over again. There are IETF drafts in this area, but one simple idea is to at least keep track of what has been pushed already on a connection. That way, you at least never push a URL more than once on a single connection.

Basically keep an unordered map, per session.

**Comment by maskit** [comment]
There are two cases that a client already have pushed content. a) The content was previously requested / pushed on previous sessions. b) The content was previously requested / pushed on the same (current) session.

For a), there are some proposed ways to avoid unnecessary push, which are not implemented on neither major browsers nor ATS. There's nothing we can do right now.

For b), we may be able to have the URL list / map, but I'm not sure whether we should do that. The content might be stale and be needed to reload. If you see ATS as a client (of an origin server), ATS should lookup own cache, and if it was stale, ATS should request (and push) the content. If it was fresh, then  ATS don't have to anything (just ignore the "preload"). Maybe, this is the way we should take. Not sure what the "preload" spec says.

We can check cache and manage the URLs on a plugin but I think we should do them on ATS core side when Server Push API has been called (if we do this), because it's not "preload" specific.

@calavera So, it would be great if you could add some logic (not limited to the above) to avoid redundant push on either the plugin or ATS core, but I'm fine without it.

Anyway, I'll review the source code later :)

**Comment by zwoop** [comment]
Adding a "TTL" to the hash wouldn't be too difficult, no? It just seems like that always pushing every asset on a page every time is a huge waste (and likely the UA can't reset those streams fast enough). I imagine the typical use case here is to push the static assets on a page, and those assets in many cases are all the same while navigating the site.

That much said, I agree, we can't enforce this on the original author here, so it's merely a suggestion for an improvement.

**Comment by masaori335** [comment]
@calavera Cool. Please add the Apache License Version 2 header in source file and run clang-format:)

**Comment by maskit** [comment]
@zwoop I thought a simple URL list :) TTL sounds good idea, I still have to check if some spec mention about it though.

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    ClientRequest &request = transaction.getClientRequest();
+    if (request.getMethod() != HTTP_METHOD_GET) {
+      return;
+    }
+
+    Response &response = transaction.getClientResponse();
+    Headers &headers = response.getHeaders();
+
+    TSHttpTxn txnp = static_cast<TSHttpTxn>(transaction.getAtsHandle());
+    string linkHeader = headers.values("Link");
```
Comment (masaori335) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: Why don't you use `header_field_iterator Headers::find(const std::string &key)` ?

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    const Url &clientUrl = request.getPristineUrl();
+
+    if (linkHeader.empty()) {
+      TSDebug(PLUGIN_NAME, "Link header not found for url: %s", clientUrl.getUrlString().c_str());
+      return;
+    }
+
+    TSDebug(PLUGIN_NAME, "Found Link header: %s", linkHeader.c_str());
+    stringstream ss(linkHeader);
+    vector<string> links = split(linkHeader, ',');
```
Comment (masaori335) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: If you use `Headers::find(const std::string &key)` in above, you don't need split them by yourself. Something I'm missing?

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  {
+    serverPush(transaction);
+    transaction.resume();
+  }
+
+  void
+  serverPush(Transaction &transaction)
+  {
+    ClientRequest &request = transaction.getClientRequest();
+    if (request.getMethod() != HTTP_METHOD_GET) {
```
Comment (masaori335) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: @maskit Is there no TS APIs to detect client protocol like `cqpv`?

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      return;
+    }
+
+    TSDebug(PLUGIN_NAME, "Found Link header: %s", linkHeader.c_str());
+    stringstream ss(linkHeader);
+    vector<string> links = split(linkHeader, ',');
+
+    for (const string& link : links) {
+      TSDebug(PLUGIN_NAME, "Link: %s", link.c_str());
+      if (link.find("nopush") != string::npos) {
```
Comment (maskit) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: This check should be a bit more strict. "nopush" might be found in the href attribute. Also, other strings (e.g. "volcanopush", "nopushpin") will pass this check. Probably we should find " nopush" ("no push" with a leading whitespace) and check if the following character is ';' or end of string.

File: plugins/experimental/server_push_preload/server_push_preload.cc (hunk 3 above)
Comment (maskit) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: That sounds good but there's no TS API for it as far as I know.

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+#include <ts/experimental.h>
+#include <atscppapi/GlobalPlugin.h>
+#include <atscppapi/utils.h>
+
+#define PLUGIN_NAME "server_push_preload"
+
+using namespace std;
+using namespace atscppapi;
+
+static regex linkRegexp("<([^>]+)>;.*rel=preload");
```
Comment (maskit) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: This will match "xrel=preloadx". At least the attribute name should be checked strictly because attribute name is extensible (Non-standard unknown names have to be expected).

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    }
+    return tokens;
+  }
+};
+
+void
+TSPluginInit(int argc ATSCPPAPI_UNUSED, const char *argv[] ATSCPPAPI_UNUSED)
+{
+  TSDebug(PLUGIN_NAME, "Init");
+  RegisterGlobalPlugin("ServerPushPreloadPlugin", PLUGIN_NAME, "david.calavera@gmail.com");
```
Comment (maskit) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: I'm not 100% sure but I think we should use dev@trafficserver.apache.com here and you can claim copyright in NOTICE file if you want.
@zwoop Am I correct?

**Comment by maskit** [comment]
I created an issue (#2205) about the concern of pushing the same content repeatedly.

File: plugins/experimental/server_push_preload/server_push_preload.cc (hunk 3 above)
Comment (shukitchan) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: I have been using this one - https://docs.trafficserver.apache.org/en/latest/developer-guide/api/functions/TSClientProtocolStack.en.html

That should work. right?

Comment (masaori335) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: Right!

File: plugins/experimental/server_push_preload/server_push_preload.cc (hunk 6 above)
Comment (zwoop) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: +1 on dev@trafficserver

**Comment by calavera** [comment]
thanks for the review, I'll address your comments soon 🙌

**Comment by scw00** [comment]
[approve ci]

**Comment by maskit** [comment]
[approve ci autest]

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  void
+  serverPush(Transaction &transaction)
+  {
+    TSHttpTxn txnp = static_cast<TSHttpTxn>(transaction.getAtsHandle());
+    if (TSHttpTxnClientProtocolStackContains(txnp, "h2") == nullptr) {
+      return;
+    }
+
+    ClientRequest &request = transaction.getClientRequest();
+    if (request.getMethod() != HTTP_METHOD_GET) {
```
Comment (maskit) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: Should we limit methods?

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  unordered_map<string, bool>
+  split(const string &params, char delim)
+  {
+    stringstream ss(params);
+    string s;
+    unordered_map<string, bool> tokens;
+    while (getline(ss, s, delim)) {
+      s.erase(find_if(s.rbegin(), s.rend(), not1(std::ptr_fun<int, int>(isspace))).base(), s.end()); // trim left
+      s.erase(s.begin(), find_if(s.begin(), s.end(), not1(std::ptr_fun<int, int>(isspace))));        // trim right
+      tokens[s] = true;
```
Comment (masaori335) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: Just curious, where this value is used? `std::set` is enough?

File: plugins/experimental/server_push_preload/server_push_preload.cc (hunk 7 above)
Comment (calavera) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: @maskit I don't really have any reason to limit it, I'll remove this check.

File: plugins/experimental/server_push_preload/server_push_preload.cc (hunk 8 above)
Comment (calavera) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: Set is enough, I updated the code. Sorry I'm too used to the lack of Set in Golang 😭

**Comment by masaori335** [comment]
[approve ci]

File: plugins/experimental/server_push_preload/server_push_preload.cc [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+          TSDebug(PLUGIN_NAME, "Push preloaded content: %s", lu.c_str());
+          TSHttpTxnServerPush(txnp, lu.c_str(), lu.length());
+        } else {
+          TSDebug(PLUGIN_NAME, "No match found for link header: %s", link.c_str());
+        }
+      }
+    }
+  }
+
+  set<string, bool>
```
Comment (maskit) [inline plugins/experimental/server_push_preload/server_push_preload.cc]: Probably, the compile error is caused by this line.

**Comment by maskit** [comment]
[approve ci]

**Comment by maskit** [comment]
[approve ci]
